src/main/java/io/prometheus/wls/rest/UrlBuilder.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2020, Oracle Corporation and/or its affiliates.
+ *
+ * Licensed under the Universal Permissive License v 1.0 as shown at http://oss.oracle.com/licenses/upl.
+ */
+
+package io.prometheus.wls.rest;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import javax.servlet.http.HttpServletRequest;
+
+import io.prometheus.wls.rest.domain.Protocol;
+
+public class UrlBuilder {
+
+  private static final List<Integer> successes = new ArrayList<>();
+
+  private final String host;
+  private final Protocol protocol;
+  private final Queue<Integer> ports = new PriorityQueue<>(new PreferSuccesses());
+  private int lastCandidate;
+
+  public UrlBuilder(HttpServletRequest request, Integer restPort) {
+    host = request.getLocalName();
+    protocol = Protocol.getProtocol(request);
+    Optional.ofNullable(restPort).ifPresent(ports::add);
+    addPortIfNotPresent(request.getLocalPort());
+  }
+
+  private void addPortIfNotPresent(int port) {
+    if (!ports.contains(port)) ports.add(port);
+  }
+
+  public static void clearHistory() {
+    successes.clear();
+  }
+
+  public String createUrl(String urlPattern) {
+    return protocol.format(urlPattern, host, selectPort());
+  }
+
+  private int selectPort() {
+    if (ports.isEmpty()) throw new WebClientException("No connection port is defined");
+
+    return lastCandidate = ports.peek();
```

Comment:
can multiple threads share the same UrlBuilder instance?

Is the following scenario possible?

restPort is configured but is not accessible. localPort is accessible.

Time 1) Thread1 calls createUrl() and got URL with restPort, lastCandidate set to restPort
Time 2) Thread2 calls createUrl() and also got URL with restPort, lastCandidate set to restPort
Time 3) Thread1 reports failure -> UrlBuilder removes lastCandidate, which is restPort, from ports
Time 4) Thread1 calls createUrl() and got URL with localPort, lastCandidate set to localPort
Time 5) Thread2 reports failure -> UrlBuilder remove lastCandidate, which is localPort, from ports, which is incorrect because the failure that Thread2 reports was for the URL created with restPort

Reply:
I have no tests covering that case. It's not likely because of the way the exporter is used (Prometheus calls it about once per second), but it is worth looking into ways to prevent that from happening - at the very least it should not set up a permanent problem.

Reply:
Looking again, I see that each request would get its own UrlBuilder (see LiveConfiguration.createUrlBuilder()), with its own set of candidate ports, and therefire the scenario you describe is impossible. The only thing shared is the successes collection, which is safe, as it is nothing but an optimization, allowing ports which have succeeded in the past to be tried first.

Reply:
agree. I have marked this as resolved.